lib/Test2/Harness/Collector.pm

```diff
@@ -8,6 +8,9 @@ use Carp qw/croak/;
 
 use Test2::Harness::Collector::JobDir;
 
+# Max open files handles on a process applies. Usually this is 1024 so we can't have everything open at once when we're behind.
+use constant MAX_JOBS_TO_MONITOR => 200; # Allow for multiple handles per job to be open
```

Comment:
I would prefer this be a runner option (See App::Yath::Options::Runner)

Reply:
And I think the default should be undefined, with no limiting happening unless it is defined.

Reply:
I very much disagree that default should be undefined. The real question is... if I have 2000 tests to parse, why should I open the file handles for all of them at once? Also what is the advantage to this being configurable? When would I want to bump this number and how would it be faster?

Reply:
There are 2 main reasons:

1) The philosophy of Test and Specially Yath is to make EVERYTHING configurable. You can change the number of jobs, you can change if it uses TAP vs Stream, etc. This factors into it. People may decide they want to change the max number of open handles instead of limiting the collector. Their reasons may not be obvious to us.

2) When parsing STDOUT/STDERR and/or TAP without the Test2 event stream, timing data in the log is dependent on when the collector reads data, not when the data was written by the test. In your solution here a backed up collector will avoid reading new jobs until the backed-up ones are completely read, this directly goes against the design of the collector, which limits how much it reads from each job per iteration so that it can jump and read from all jobs. In your solution a handful of noisy jobs can offset timing data and display of other jobs. That is not necessarily wrong, but should be a choice, not a baked in design decision.

-- Additional Ideas (not quick, so probably not useful here, but worth mentioning anyway) --

If I were in this position I would work on making the tests that cause the collector to get backed-up less noisy, if their output is entirely important data (test results) then that is obviously not an option.

We might also consider making the collector more parallel (IE have 1 collector for every N job slots, also configurable) to prevent the backup, that would help with system/user filehandle limits a little bit by having less backups, but may not solve your problem. It would absolutely help with process fd limits.

Reply:
As far as the collector parsing each job in parallel, I totally support this idea. It's just too much scope to fix right now.

Reply:
I would suggest at at this point cPanel is the only group using the collector to this level. Anyone who lets the collector get to the point of there being more than 500 jobs to parse in one loop is going to crash. We picked 200 semi-arbitrarily but I think most people who hit this would be surprised. Which is why I suggest a default limit. I could go as high as 400 for the default but I think it's a mistake not to set it. I doubt very much anyone but us has hit this limit.